Describe Force progress production (#462)
aeternity/protocol: node/api/channels_api_usage.md

`round` will replace the other and all force-progressed states that had been
based on the older state.

## Force progress

The biggest strength of State Channels lies in having fast and cheap off-chain
contract execution. This imposes a risk, though: if the other party suddenly

> becomes non-cooperative or simply missing - a new off-chain state can not be

Reviewer: ```suggestion
becomes non-cooperative or simply missing, a new off-chain state can not be
```

produced. This is where the `channel_force_progress_tx` transaction comes in.

> It allows any of the participant to unilaterally execute off-chain contract

Reviewer: ```suggestion
It allows any of the participants to unilaterally execute an off-chain contract
```

on-chain. This produces the next State Channel off-chain state on-chain.

> Channel force progress transaction is based on the latest co-authenticated

Reviewer: ```suggestion
A channel force progress transaction is based on the latest co-authenticated
```

state. It provides off-chain state trees on-chain. They contain the contract
to be executed and all the context needed for the execution itself. This
breaks the assumption of off-chain privacy. If successful, the
`channel_force_progress_tx` transaction produces on-chain the next off-chain
state. That's why it also contains the next `round` and the next `state_hash`.
The latter is the result of applying the off-chain contract call to the
off-chain state trees.

> Force progress transaction can be used while the channel is being closed or

Reviewer: ```suggestion
A force progress transaction can be used while the channel is being closed or
```

> while it is still open. Assumption is if one participant refuses to cooperate,

Reviewer: ```suggestion
while it is still open. The assumption is that if one participant refuses to cooperate,
```

> if the other produces the next forced progress state on-chain - there is no

Reviewer: ```suggestion
if the other produces the next forced progress state on-chain, there is no
```

going back. From then on they could continue either cooperating or they could

> close the channel. In both cases - they use the on-chain produced off-chain

Reviewer: ```suggestion
close the channel. In both cases, they use the on-chain produced off-chain
```

state.

> The `channel_force_progress_tx` transaction could be a challange to get right.

Reviewer: ```suggestion
The `channel_force_progress_tx` transaction could be a challenge to get right.
```

The FSM handles this for the client and from a client's perspective it looks

> like the off-chain contract call. Only difference is that with

Reviewer: ```suggestion
like the off-chain contract call. The only difference is that with
```

> `channel_force_progress_tx` the `gas_price` is mandatorary. It is being used

Reviewer: ```suggestion
`channel_force_progress_tx` the `gas_price` is mandatory. It is used
```

both in the contract call execution and the transaction's `fee` computation.

#### Forcer inittiates a forced progress

Any participant can initiate a forced progress by:

```javascript
{
  "jsonrpc": "2.0",
  "method": "channels.force_progress",
  "params": {
    "abi_version":1,
    "amount":10,
    "call_data":"cb_AAAAA...",
    "contract_id":"ct_5XjcY...",
    "gas_price":1000000000
  }
}
```

> Except of the `gas_price`, all other `params` are corresponding to those in a

Reviewer: ```suggestion
Except of the `gas_price`, all other `params` correspond to those in a
```

> `call_contract` off-chain update. Reasoning is quite trivial: if the other

Reviewer: ```suggestion
`call_contract` off-chain update. The reasoning is quite trivial: if the other
```

```javascript
{ 
   "jsonrpc":"2.0",
   "method":"channels.sign.force_progress_tx",
   "params":{ 
      "channel_id":"ch_2FdiLKkRUdPw4oTRbB6i3M6pquogzWLABQjU373hizDbnD8gGC",
      "data":{ 
         "signed_tx":"tx_+Qi9CwHAuQ....",
         "updates":[ 
            { 
               "abi_version":1,
               "amount":10,
               "call_data":"cb_AAAAAAA...",
               "call_stack":[ 

               ],
               "caller_id":"ak_Vu1cG...",
               "contract_id":"ct_55C...",
               "gas":1000000,
               "gas_price":1000000000,
               "op":"OffChainCallContract"
            }
         ]
      }
   },
   "version":1
}
```

Note that the `updates` that comes with the `channel_force_progress_tx` is the
same as it would be if it were an off-chain `call_contract`.

The forcer is to decode the transaction, inspect its contents, authenticate
it, encode it and then post it back via a WebSocket message:

```javascript
{
  "jsonrpc": "2.0",
  "method": "channels.force_progress_sign",
  "params": {
    "signed_tx": "tx_+QFxCwH4QrhA..."
  }
}
```

The FSM is to check the transaction and its authentication and then post it

> from forcer's behalf on-chain. Once it detects it being included

Reviewer: ```suggestion
on the forcer's behalf on-chain. Once it detects it being included
```

> on-chain, it would report it:

Reviewer: ```suggestion
on-chain, it reports it:
```

```javascript
{
  "jsonrpc": "2.0",
  "method": "channels.on_chain_tx",
  "params": {
    "channel_id": "ch_2eiGsAvt...",
    "data": {
      "info": "channel_changed",
      "tx": "tx_+QFxCwH4QrhA...",
      "type": "channel_force_progress_tx"
    }
  },
  "version": 1
}
```

> The other participant's FSM will also notify its client that it had seen a new

Reviewer: ```suggestion
The other participant's FSM will also notify its client that it has seen a new
```

> transaction changing the channel on-chain being included in a microblock.

Reviewer: ```suggestion
transaction in a microblock, changing the channel on-chain.
```

An interesting edge case would be one participant producing a force progress based on

> a state that was valid at some point of time but not the latest one. This is

Reviewer: ```suggestion
a state that was valid at some point of time but is not the latest one. This is
```
